docs/rfcs/033-pageserver-postgres-bundle.md

```diff
@@ -0,0 +1,139 @@
+# Postgres Bundle for Pageserver
+
+Created on 2024-06-17
+
+## Summary
+
+This RFC defines the responsibilities of Compute and Storage team regarding the
+build & deployment of the Postgres code that Pageserver must run
+(`initdb`, `postgres --wal-redo`).
+
+## Motivation
+
+Pageserver has to run Postgres binaries to do its job, specifically
+
+* `initdb`
+* `postgres --wal-redo` mode
+
+Currently there is **no clear ownership** of
+* how these binaries are built
+  * including, critically, dynamic linkage against other libraries such as `libicu`
+* what build of the binaries ends up running on Pageservers
+* how the binaries and runtime dependencies (e.g., shared libraries) are delivered to Pageservers
+
+Further, these binaries have dependencies (e.g., libicu) which
+1. prevent the Storage team from switching Pageserver distro and/or version, and
+2. some dependencies impact compatibility between Storage and Compute (e.g., [libicu version impacts collation incompatibilty](https://github.com/neondatabase/neon/pull/8074))
+3. some dependencies can cause database corruption if updated carelessly (locale => libc)
+
+## Why Is This Worth Solving
+
+1. Clearly defined ownership generally boosts execution speed & bug triage.
+   * Example for why execution speed matters: CVE in dependency => who takes care of patching & updating.
+2. Centralize understanding of risks involved with some dependencies.
+   Currently, there is no team clearly responsible for assessing / tracking the risks. As a reminder from previous section, these are
+   * runtime incompatibilities
+   * database corruption
+
+Also, it is an unlock for additional future value, see "Future Work" section.
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+Pageserver (neon.git)
+Compute (neon.git)
+Deployment process (aws.git)
```

Comment:
One additional component that is flying under the radar (and IMO should be adapted rather than treated as a first-class component): the **neon_local setup**.

neon_local (including our Python regression tests) use the same Postgres build for Computes and Pageserver.
However, they don't _need_ the same build config.
For example, the Pageserver doesn't need libicu nor openssl for initdb and walredo.

It might make sense to build Postgres twice, once for Compute (dynamically linked, etc) and once for Pageserver, statically linked (Variant 1 above).

Reply:
> For example, the Pageserver doesn't need libicu nor openssl for initdb and walredo.

I actually not 100% sure about not needing libicu in wal-redo. I think wal redo functions usually do not need e.g. value comparison (which would depend on libicu/locale for strings), but I don't know how to prove it. 

@hlinnaka @MMeent will anything break in wal-redo without icu? If yes we could build a separate fully static binary for it.